File: cpp/763-partition-labels.cpp
class Solution {
public:
    vector<int> partitionLabels(string s) {
        unordered_map<char, int> hashmap;
        for (int i = 0; i < s.length(); i++) {
            if (hashmap.find(s[i]) == hashmap.end()) {
                hashmap.insert({ s[i], i });
            } else {
                hashmap[s[i]] = i;
            }
        }

        vector<int> res;
        int size = 0;
        int end = 0;
        for (int i = 0; i < s.length(); i++) {
            end = max(end, hashmap[s[i]]);
            size++;

            if (end == i) {
                res.push_back(size);
                size = 0;
            }
        }
        return res;
    }
};

File: rust/763-partition-labels.rs
use std::collections::HashMap;
use std::cmp;

impl Solution {
    pub fn partition_labels(s: String) -> Vec<i32> {
        let mut hashmap: HashMap<char, usize> = HashMap::new();
        for (i, c) in s.chars().enumerate() {
            hashmap.insert(c, i);
        }
        let mut res: Vec<i32> = Vec::new();
        let mut size = 0;
        let mut end = 0;
        for (i, c) in s.chars().enumerate() {
            end = cmp::max(end, *hashmap.get(&c).unwrap());
            size += 1;

            if end == i {
                res.push(size as i32);
                size = 0;
            }
        }
        res        
    }
}